llama-index-integrations/readers/llama-index-readers-file/llama_index/readers/file/unstructured/base.py

```diff
@@ -15,98 +20,186 @@
-
-                if extra_info is not None:
-                    metadata.update(extra_info)
-
-                metadata["filename"] = str(file)
-                docs.append(Document(text=node.text, extra_info=metadata))
+    def _create_documents(
+        self,
+        elements: list,
+        document_kwargs: Optional[dict],
+        extra_info: Optional[dict],
+        split_documents: Optional[bool],
+        deterministic_ids: Optional[bool],
+        excluded_metadata_keys: Optional[list[str]],
+    ) -> list[Document]:
+        """Create documents from partitioned elements.
+
+        Args:
+            elements (List): List of partitioned elements.
+            document_kwargs (Optional[Dict]): Additional arguments for document creation.
+            extra_info (Optional[Dict]): Extra information to add to the document metadata.
+            split_documents (Optional[bool]): Whether to split the documents.
+            deterministic_ids (Optional[bool]): Whether to generate deterministic IDs.
+            excluded_metadata_keys (Optional[List[str]]): Keys to exclude from the metadata.
+
+        Returns:
+            List[Document]: List of parsed documents.
+        """
+        document_kwargs = document_kwargs.copy() if document_kwargs else {}
+        docs: list[Document] = []
 
+        if split_documents:
+            for sequence_number, element in enumerate(elements):
+                kwargs = document_kwargs.copy()
+                kwargs["text"] = element.text
+
+                excluded_keys = set(
+                    excluded_metadata_keys or self.excluded_metadata_keys
+                )
+                metadata = extra_info.copy() if extra_info else {}
+                for key, value in element.metadata.to_dict().items():
+                    if key not in excluded_keys:
+                        metadata[key] = (
+                            value
+                            if isinstance(value, self.allowed_metadata_types)
+                            else json.dumps(value)
+                        )
+
+                kwargs["extra_info"] = metadata
+
+                if deterministic_ids:
+                    kwargs["doc_id"] = element.id_to_hash(sequence_number)
+
+                docs.append(Document(**kwargs))
         else:
             text_chunks = [" ".join(str(el).split()) for el in elements]
+            text = "\n\n".join(text_chunks)
+
+            kwargs = document_kwargs.copy()
+            kwargs["text"] = text
+
+            metadata = extra_info.copy() if extra_info else {}
+
+            if len(elements) > 0:
+                filename = elements[0].metadata.filename
+                metadata["filename"] = filename
+
+                if deterministic_ids:
+                    from hashlib import sha256
 
-            metadata = {}
+                    data = f"{filename}{text}"
+                    kwargs["doc_id"] = sha256(data.encode()).hexdigest()[:32]
```

Comment:
wouldn't a more consistent ID be the file path? 

Because now, when the content changes, the ID will be new (and there is no way to ground the comparison to the original document if its already inserted)

Reply:
This is the same method of hashing that unstructured is using in `element.id_to_hash(sequence_number)`, except we don't have a sequence number, which doesn't apply anyway because we're making the whole text the `Document` rather than a specific chunk.
I think that is the point, if the content changes, you want it to change such that it's considered a different document than the previous?

Reply:
Without an consistent ID, we can't go back and update the original document/chunks

Reply:
There needs to be some anchor for comparing to existing data, which is commonly the doc id

Reply:
If it changes every time, upserts will not happen properly

Reply:
seems like to me we can use the filename/path ?

Reply:
I also see a deterministic_ids list 😅 Sounds useful